src/tools/idlc/src/generator.c

```diff
@@ -395,12 +396,49 @@ idlc_generate(const idl_pstate_t *pstate, const idlc_generator_config_t *config)
   memset(&generator, 0, sizeof(generator));
   generator.path = file;
 
+  if(config->base_dir) {
+    char* rel_path = NULL;
+    char* old_dir = dir;
+    // If we run into a path error we can still recover and use existing dir
+    if(idl_relative_path(config->base_dir, pstate->sources->path->name, &rel_path) == IDLC_NO_MEMORY)
+      goto err_mkpath;
+    // If root is not a parent of file path, or there was an error resolving rel path
+    // TODO maybe find a common directory between the two and build from there?
+    if(rel_path == NULL || strncmp(rel_path, "..", 2) == 0)
+      goto use_dir_path;
+    dir = NULL;
+    size_t print_len = strlen(rel_path);
+    if(sep)
+      print_len -= strlen(sep);
+    if(!(dir = idl_strndup(rel_path, print_len))){
+      dir = old_dir;
+      goto use_dir_path;
+    }
+    if(old_dir && old_dir[0] != '\0')
```

Comment:
I'd appreciate it if you would help me understand why the `old_dir[0]!=0` is needed.